File: examples/poodle-surf/android-java/app/src/main/java/org/diez/examples/poodlesurfjava/MainActivity.java
package org.diez.examples.poodlesurfjava;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import org.diez.poodleSurf.*;

public class MainActivity extends AppCompatActivity {
    private DesignSystem diez;
    private ModelMocks mocks;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        new Diez<>(new DesignSystem(), findViewById(android.R.id.content)).attach((DesignSystem component) -> {
            runOnUiThread(() -> {
                diez = component;
                onDiezUpdated();
            });

            return null;
        });

        new Diez<>(new ModelMocks(), findViewById(android.R.id.content)).attach((ModelMocks component) -> {
            runOnUiThread(() -> {
                mocks = component;
                onMocksUpdated();
            });

            return null;
        });
    }

    private void onDiezUpdated () {
        System.out.println("======== DS updated ===========");
        System.out.println(diez);

    }

    private void onMocksUpdated () {
        System.out.println("======== Mocks updated ===========");
        System.out.println(mocks);
    }
}

File: packages/compiler/playground/android/HelloMyStateBag/app/src/main/java/ai/haiku/hellomystatebag/MainActivity.kt
package ai.haiku.hellomystatebag

import ai.haiku.diez.puente.Diez
import ai.haiku.diez.components.MyStateBag
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.MotionEvent
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    val diez = Diez(MyStateBag())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        diez.attach(layout, fun(component) {
            runOnUiThread {
                text.text = component.copy
                component.textStyle.setTextStyle(text)
                component.image.setBackground(view)
            }
        })
        // FIXME: Note how we must to do this *after* we've called diez.attach() at least once.
        // This is because the environment hasn't been correctly initialized yet.
        runOnUiThread {
            diez.component.haiku.embedHaiku(haikuSlot)
            diez.component.svg.embedSvg(ratSlot)
            diez.component.lottie.embedLottie(lottieSlot)
        }
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        if (e.action == MotionEvent.ACTION_DOWN) {
            diez.component.tap()
        }
        return true
    }
}
